docs/docs/typelevel.md

```diff
@@ -0,0 +1,532 @@
+apart from blocks with normal case clauses which define partial
+functions and therefore can only appear in normal code.  Every query
+conditional is evaluated by the macro interpreter. The macro
+interpreter evaluates each case of a query conditional by performing
+an implicit search for the given type pattern. If the search succeeds,
+it continues with evaluating the guard, if there is one, followed by
+the right-hand side, otherwise it continues with the next case.
+
+Type patterns may bind type variables. For instance,
+`N =:= S[type N1]` binds the type variable `N1`. Bound type variables are
+instantiated by the implicit search. They are visible in the rest of
+the case.
+
+In general, a type name in an implicit condition is treated as a bound
+variable if it is prefixed by `type`.
+
+_Aside:_ We should introduce the same convention for types in normal
+patterns. Currently type identifiers in patterns are treated as bound
+if they start with a lower-case letter and as free otherwise. This
+mirrors the convention for names, but is not very visible. Requiring
+`type` prefixes instead would make programs clearer.
+
+In last clause of the code above, the macro interpreter would perform
+an implicit search for the type `N =:= S[type N1]` where `N` is the
+actual type passed to `toInt` and `N1` is a type variable bound by the
+search.  If the search succeeds, it returns one plus the result of
+evaluating `toInt[N1]`. If the search fails, it results in compilation
+failure.
+
+The result of `toInt[S[S[Z]]]` would hence be the integer `1 + (1 +
+0)`, which can be constant-folded to `2`. Alternatvely, we could drop
+the use of quotes and rely on lifting to go from `Int` to
+`Expr[Int]`. So the following would also be correct and would return a
+constant without relying on constant folding:
+
+    inline def toInt[N <: Nat]: Int = ~{
+      case N =: Z => 0
+      case N =:= S[type N1] => 1 + toInt[N1]
+    }
+
+### Example: HLists
+
+Consider the standard definition of `HList`s:
+
+    class HList
+
+    case class HCons[X, Xs <: HList](hd: X, tl: Xs) extends HList
+    case object HNil extends HList
+
+A type-level length function on HLists can be written as follows:
+
+    type Length[Xs <: HList] <: Nat = {
+      case Xs =:= HNil => Z
+      case Xs =:= HCons[type X, type Xs1] => S[Length[Xs1]]
+    }
+
+If we don't want to introduce the type variable `X`, which is unused, we could
+also formulate the type pattern with `<:<` instead of `=:=`:
+
+    type Length[Xs <: HList] <: Nat = {
+      case Xs <:< HNil => Z
+      case Xs <:< HCons[_, type Xs1] => S[Length[Xs1]]
+    }
+
+Here's a `Concat` type with associated `concat` method:
+
+    type Concat[Xs <: HList, Ys <: HList] <: HList = {
+      case Xs =:= HNil => Ys
+      case Xs =:= HCons[type X, type Xs1] => HCons[X, Concat[Xs1, Ys]]
+    }
+
+    inline def concat[Xs <: HList, Ys <: HList](xs: Xs, ys: Ys): Concat[Xs, Ys] = ~ {
```

Comment:
Making `concat` inline allows using the rules for typechecking query conditionals, but it does not seem expressive enough. Usually type-level programming lets me define a proper *runtime* function `concat` and typecheck it once and for all; if we can’t do that, it seems we forbid lots of applications of `HList`. And to write a runtime `concat` we have to either do contract-checking for `concat` at the actual runtime, and define erasure for types like `Concat[Xs, Ys]` (so that type errors might manifest not just while running macros but while running functions), or to introduce symbolic evaluation.

Reply:
I think this point illustrates that we have different expectations. Defining a runtime `concat` is possible only if the _representation_ of types is homogeneous, so `HLists` or tuples are represented as normal lists at runtime. But there's no way we can or want to do this! So compile-time is the only option.

I think the root issue is that you are coming from a world that is not very applicable to our constraints. We want to achieve similar expressiveness as dependently typed languages, but in a language where efficiency matters, representations have to be specialized, and everything is subject to suptyping. This means we need to be inspired by the likes of C++ or D as well as the sources you cite.

Also, the whole proposal really makes sense only in conjunction with the new meta-programming proposal:

    http://dotty.epfl.ch/docs/reference/symmetric-meta-programming.html

That's where the synergy comes from! Seeing this as an isolated approach to typelevel programming will lead you astray, IMO.

Reply:
Thanks for linking to http://dotty.epfl.ch/docs/reference/symmetric-meta-programming.html, I’ll catch up and sorry if I jumped ahead too much. I also had to look at https://github.com/lampepfl/dotty/pull/2199 to understand the plans for tuples—but there seems to be a common `Tuple` type?

Overall, I’d just like to (help) understand and clarify limitations and tradeoffs of proposals like this, and get to some sweet spot — that is, try to remove any limitations that don’t buy something in exchange — not impose my preference about a certain sweet spot. Of course I’ll compare with related work I’m familiar with, even though that’s not perfect.

Here, you talk about usual type-level programming and standard HLists. So just to understand: is this meant to subsume existing Scala approaches, as the introduction suggests, or is this specialized for a few common scenarios? Current Scala designs allow a runtime `concat` in Scala today.* Now that you mention C++, maybe I understand better what you have in mind; I might later propose edits to the introduction to clarify what you mean. Yet, in C++ having to specialize everything can cause excessive code bloat, so I’d be afraid if we have to go there.

And is a runtime `concat` impossible also if I use some other data structure where I make different choices?

*I checked again https://apocalisp.wordpress.com/2010/07/08/type-level-programming-in-scala-part-6b-hlist%C2%A0folds/ and there are runtime operations—where the type is computed statically. Copy-pasted sample:

```scala
// construct a heterogeneous list of length 3 and type
//  Int :: String :: List[Char] :: HNil
val a = 3 :: "ai4" :: List('r','H') :: HNil
 
// construct a heterogeneous list of length 4 and type
//  Char :: Int :: Char :: String :: HNil
val b = '3' :: 2 :: 'j' :: "sdfh" :: HNil
 
   // append the two HLists
val ab = a ::: b
```

Reply:
@Blaisorblade please take a look at this code sample, I hope you'll find it useful: https://scastie.scala-lang.org/senia-psm/zqIQMbtgTk6YQj6nlY7Ppw/12

In object `concat_using_=:=` you can find code that works in current version of dotty, but looks similar to `concat` implementation from this PR.

In part "using type macto" I describe my understanding of possible runtime representations of `concat(a, b)`.

Also you may take a look at `manual_concatter` for runtime representation of `a ::: b`.

Reply:
Here's a tongue in cheek version to do a runtime `concat`:

    trait Concatenizer[Xs <: HList, Ys <: HList] {
       def apply(xs: Xs, ys: Ys): Concat[Xs, Ys]
    }

    implicit inline def concatenizer[Xs <: HList, Ys <: HList]: Concatenizer[Xs, Ys] = new {
       def apply(xs: Xs, ys: Ys) = concat(xs, ys)
    }

   def runtimeConcat[Xs <: HList, Ys <: HList](xs: Xs, ys: Ys)(implicit conc: Concatenizer[Xs, Ys]) = 
      conc.apply(xs, ys)

That's indeed a runtime concat but it defers to an implicit that does the actual work and that is synthesized at compile time. The example you were pasting is the same in spirit. And yes, that can generate _lots_ of code.

Reply:
Well, here at least the values *can* be picked at runtime, so we keep a phase distinction and specialize per shape and not per values. I’m curious how much of the shape one needs to reify at compile-time, and if staging lets us have control of this.

But I’ll have to study more closely both HList and @senia-psm’s snippet!

Reply:
What's not explicit in the proposal is when a case-defined type like `Concat[A,B]` is expanded. In `trait Concatenizer`, it can clearly not be expanded because we don't know `Xs` and `Ys` yet. This seems to imply that expansion of a case-defined type `T` would happen only while checking that a term `t` has type `T`, as in the definition of `apply` in `concatenizer` where we check that `concat(xs, ys)` has type `Concat[Xs, Ys]`. Is that right?

Reply:
The status of the `Concat[Xs, Ys]` type in `trait Concatenizer` in relation with logic programming makes it appear that this approach is somewhat related to type projections.

A case-defined type such as `Concat[A,B]` can be encoded as `type Concat[A,B] { type Out }` (as in @senia-psm's code) where `Out` is the result of the type-level computation and where each case of the case-defined type is an implicit definition (potentially mutually-recursive with the other cases). Then `Concat[Xs, Ys]` in `trait Concatenizer` has to be `Concat[Xs, Ys] # Out` because there is no implicit value in scope at this point.

So this seems to regain some of the extra flexibility that we had with type projections (which was arguably one reason why they were so pervasive in type-level programming). Hopefully this new approach does not have the same soundness problems as type projections.

Reply:
Usage of `Concat` in trait looks really confusing for me. I would expect something like this:
```
trait Concatenizer[Xs <: HList, Ys <: HList] {
  type Out <: HList
  def apply(xs: Xs, ys: Ys): Out
}
implicit inline def concatenizer[Xs <: HList,
                             Ys <: HList]: Concatenizer[Xs, Ys] { type Out = Concat[Xs, Ys] } = new {
   def apply(xs: Xs, ys: Ys): Concat[Xs, Ys] = concat(xs, ys)
}
```

Reply:
Let's forget about the example, which was anyway tongue-in-cheek.